src/segmentation.rs

```diff
@@ -0,0 +1,102 @@
+//! Segmenter trait and default implementations.
+//!
+//! Segmentation is how strings are split into tokens for comparison.
+//! For example, two strings that *appear* identical might have different byte-level representations.
+//!
+//! Take `ä` and `ä`. Visually, these should be identical. However, the former
+//! is Unicode character [ä (U+00E4)](https://www.compart.com/en/unicode/U+00E4)
+//! while the latter is two adjacent Unicode characters [a (U+0061)](https://www.compart.com/en/unicode/U+0061)
+//! and [◌̈ (U+0308)](https://www.compart.com/en/unicode/U+0308).
+//!
+//! Depending on the [Segmenter] used, comparing these strings will return different results.
+//!
+//! ```
+//! # use fuzzywuzzy::segmentation::{Segmenter, ByteSegmenter};
+//! // U+00E4
+//! assert_eq!(ByteSegmenter.segment("ä"), vec![0xc3u8, 0xa4u8]);
+//! // U+0061 + U+0308
+//! assert_eq!(ByteSegmenter.segment("ä"), vec![0x61u8, 0xccu8, 0x88u8]);
+//! ```
+//! Given this segmentation, we would expect a comparison to return 0% similarity because every byte is different!
+//!
+//! However, even with more advanced segmentation strategies like [code point segmentation](CodePointSegmenter)
+//! or [grapheme segmentation](GraphemeSegmenter), these will still have 0% similarity by any comparison algorithm.
+//!
+//! ```
+//! # use fuzzywuzzy::segmentation::{Segmenter, CodePointSegmenter, GraphemeSegmenter};
+//! // U+00E4
+//! assert_eq!(CodePointSegmenter.segment("ä"), vec!['ä']);
+//! assert_eq!(GraphemeSegmenter.segment("ä"), vec!["ä"]);
+//! // U+0061 + U+0308
+//! assert_eq!(CodePointSegmenter.segment("ä"), vec!['a', '\u{0308}']);
+//! assert_eq!(GraphemeSegmenter.segment("ä"), vec!["ä"]);
+//! // U+00E4 vs. U+0061 + U+0308
+//! assert_ne!(GraphemeSegmenter.segment("ä"), GraphemeSegmenter.segment("ä"));
+//! ```
+//!
+//! In order to usefully compare strings like these, [normalization][crate::normalization] must be done prior to segmentation.
+
+/// Represents a strategy for segmenting a string into units for comparison.
+///
+/// The trait is also implemented for functions matching the signature of the `segment` method.
+pub trait Segmenter<'a> {
+    /// The type of the unit of comparison this strategy operates on.
+    type Output: 'a + Eq;
+    /// Produces units of comparison from a string according to the segmentation strategy.
+    fn segment(&self, s: &'a str) -> Vec<Self::Output>;
+}
+
+impl<'a, F: Fn(&str) -> Vec<T>, T: 'a + Eq> Segmenter<'a> for F {
```

Comment:
Could you provide an example usage of this trait? I haven't seen `impl`s on function objects before.

Reply:
it works for closures and normal functions. its nifty. i'll add an example.